File: python/src/linodemcp/profiles/validator.py
"""Phase 6.4b token-scope validator.

Mirrors ``go/internal/profiles/validator.go``. Pure orchestration:
fetch /profile, decide PAT vs OAuth from the Scopes string, fetch
/profile/grants on the OAuth branch, then run the scope comparison.
Returns a structured result; the caller (Server.validate_scopes /
main.py) decides policy.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum
from typing import TYPE_CHECKING, Any, Protocol

from linodemcp.genpb.linode.mcp.v1.profile_pb2 import (
    ProfileGetInput,
    ProfileGrantsGetInput,
)
from linodemcp.linode import parse_grants, parse_profile
from linodemcp.linode.routes import tool_of
from linodemcp.profiles.scopecheck import (
    compare_scopes,
    flatten_grants,
    parse_pat_scopes,
)

if TYPE_CHECKING:
    from collections.abc import Sequence

    from linodemcp.linode import Profile
    from linodemcp.profiles.profile import Profile as ProfileModel
    from linodemcp.profiles.scopecheck import ScopeComparison


def profile_is_elevated(profile: ProfileModel) -> bool:
    """Return True if the profile permits any mutating tool.

    Reads the capability-derived ``elevated`` flag, not scope suffixes:
    the API documents ``:read_write`` scopes on several read-only routes
    (kubeconfig, managed contacts, instance interfaces), so a read-only
    profile's scope union can legitimately contain write scopes without
    the profile being able to mutate anything. The missing-token policy
    uses this to decide whether to fail load (elevated) or
    warn-and-continue (read-only).

    Profiles with no allowed tools are NOT elevated; spec behavior is
    that such profiles can start without a token.
    """
    return profile.elevated


class TokenNotConfiguredError(Exception):
    """Raised from ``Server.validate_scopes`` when the active environment
    has no Linode token set. The caller (main) decides what to do:
    read-only profiles warn and continue, elevated profiles fail to
    start.
    """


class TokenKind(IntEnum):
    """Classifies the active token as PAT vs OAuth.

    Personal access tokens carry their scope string directly on the
    ``/profile`` response; OAuth tokens leave it empty and require a
    second call to ``/profile/grants``. The validator picks the path
    based on what ``/profile`` returns; consumers use the kind for
    logging and audit.
    """

    Unknown = 0
    PAT = 1
    OAuth = 2


class TokenInspector(Protocol):
    """The client surface ``validate_scopes`` reads through: the
    contract-routed JSON read ``RetryableClient`` provides.

    Tests inject a stub keyed by tool so the validator stays network-free.
    Async because the production client is async.
    """

    async def route_raw(self, tool: str, *values: object) -> Any: ...


@dataclass(frozen=True)
class ScopeValidationResult:
    """The validator's structured return value.

    The caller decides what to do: ``missing`` is always a hard fail at
    load time; ``excess`` is a warn by default and a fail under strict
    mode. The profile is preserved on the result so callers can log
    the username/restricted flag alongside the comparison.
    """

    kind: TokenKind
    actual_scopes: tuple[str, ...]
    comparison: ScopeComparison
    profile: Profile


class ProfileFetchError(Exception):
    """Raised when the underlying GET /profile call fails.

    Wraps the original exception so callers can match this class to
    distinguish a network/API failure from a scope mismatch (which is
    reported via the comparison, not an exception).
    """


class GrantsFetchError(Exception):
    """Raised when the OAuth-path GET /profile/grants call fails."""


async def validate_scopes(
    inspector: TokenInspector,
    required: Sequence[str],
) -> ScopeValidationResult:
    """Inspect a token's scopes and diff against the profile's required set.

    PAT path: ``Profile.scopes`` is non-empty, parse it via
    ``parse_pat_scopes`` and skip ``/profile/grants``. OAuth path:
    empty ``Profile.scopes`` triggers a grants fetch and
    ``flatten_grants`` produces the actual set.

    Policy decisions live in the caller: this function reports facts
    only. A comparison with ``missing`` entries is a load-time failure
    under the spec, but this function returns the result normally so
    callers can inspect ``missing`` and ``excess`` together.

    Each read names the tool through its generated input message rather than
    spelling the name here, so the route stays declared in one place.

    Raises:
        ProfileFetchError: when the profile read fails. Wraps the original
            exception in ``__cause__``.
        GrantsFetchError: when the grants read fails on the OAuth branch.
            Wraps the original exception in ``__cause__``.
    """
    try:
        profile = parse_profile(await inspector.route_raw(tool_of(ProfileGetInput)))
    except Exception as exc:
        raise ProfileFetchError("fetch /profile failed") from exc

    if profile.scopes:
        actual = parse_pat_scopes(profile.scopes)
        return ScopeValidationResult(
            kind=TokenKind.PAT,
            actual_scopes=tuple(actual),
            comparison=compare_scopes(required, actual),
            profile=profile,
        )

    try:
        grants = parse_grants(await inspector.route_raw(tool_of(ProfileGrantsGetInput)))
    except Exception as exc:
        raise GrantsFetchError("fetch /profile/grants failed") from exc

    actual = [str(scope) for scope in flatten_grants(grants)]
    return ScopeValidationResult(
        kind=TokenKind.OAuth,
        actual_scopes=tuple(actual),
        comparison=compare_scopes(required, actual),
        profile=profile,
    )

File: go/internal/profiles/validator.go
package profiles

import (
	"context"
	"fmt"

	"github.com/chadit/LinodeMCP/internal/linode"
)

// TokenKind identifies whether the active token is a personal access
// token (scopes embedded in /profile) or an OAuth token (grants on
// /profile/grants). The validator picks the right code path based on
// what /profile returns; consumers use the kind for logging.
type TokenKind int

const (
	// TokenKindUnknown is the zero value. Set when validation has not
	// run yet, or when the /profile response was malformed.
	TokenKindUnknown TokenKind = iota
	// TokenKindPAT marks a personal access token. PATs carry their
	// scope string directly on the /profile response.
	TokenKindPAT
	// TokenKindOAuth marks an OAuth token. OAuth tokens have an empty
	// Scopes field; the validator falls back to /profile/grants.
	TokenKindOAuth
)

// String returns the token kind name for log lines and error messages.
func (k TokenKind) String() string {
	switch k {
	case TokenKindUnknown:
		return "Unknown"
	case TokenKindPAT:
		return "PAT"
	case TokenKindOAuth:
		return "OAuth"
	default:
		return "TokenKind(invalid)"
	}
}

// TokenInspector is the minimal Linode client surface ValidateScopes
// needs. The real *linode.RetryableClient satisfies this interface
// without changes; tests inject a stub so scope validation logic stays
// network-free.
type TokenInspector interface {
	GetProfile(ctx context.Context) (*linode.Profile, error)
	GetProfileGrants(ctx context.Context) (*linode.Grants, error)
}

// ScopeValidationResult is what ValidateScopes returns to its caller.
// The caller (main.go in production, tests in unit) decides what to do
// with the comparison: a Missing set is always a hard fail, an Excess
// set is a warn by default, fail under strict mode.
type ScopeValidationResult struct {
	// Kind tells whether the active token is a PAT or OAuth.
	Kind TokenKind

	// ActualScopes is the deduplicated, sorted scope set the token
	// actually carries. For PATs this comes from Profile.Scopes; for
	// OAuth it comes from FlattenGrants.
	ActualScopes []Scope

	// Comparison holds the Missing/Excess diff against the required
	// scope list ValidateScopes was given.
	Comparison ScopeComparison

	// Profile is the /profile response, for callers that want the
	// username, restricted flag, or UID for logging. Never nil on a
	// successful return.
	Profile *linode.Profile
}

// ValidateScopes inspects the token's actual scopes and returns the
// diff against required. The PAT path uses Profile.Scopes directly;
// OAuth tokens (empty Profile.Scopes) trigger a second call to
// /profile/grants and the result is flattened via FlattenGrants.
//
// Policy decisions live in the caller: this function reports facts
// only. A Comparison with Missing entries is a load-time failure under
// the spec, but ValidateScopes itself returns nil error for that case
// so callers can inspect both Missing and Excess in one place.
//
// Errors returned from this function are network/API failures only.
// The PAT path returns ErrProfileFetchFailed wrapping the underlying
// error; the OAuth path returns ErrGrantsFetchFailed.
func ValidateScopes(
	ctx context.Context,
	inspector TokenInspector,
	required []Scope,
) (*ScopeValidationResult, error) {
	profile, err := inspector.GetProfile(ctx)
	if err != nil {
		return nil, fmt.Errorf("%w: %w", ErrProfileFetchFailed, err)
	}

	if profile.Scopes != "" {
		actual := ParsePATScopes(profile.Scopes)

		return &ScopeValidationResult{
			Kind:         TokenKindPAT,
			ActualScopes: actual,
			Comparison:   CompareScopes(required, actual),
			Profile:      profile,
		}, nil
	}

	grants, err := inspector.GetProfileGrants(ctx)
	if err != nil {
		return nil, fmt.Errorf("%w: %w", ErrGrantsFetchFailed, err)
	}

	actual := FlattenGrants(grants)

	return &ScopeValidationResult{
		Kind:         TokenKindOAuth,
		ActualScopes: actual,
		Comparison:   CompareScopes(required, actual),
		Profile:      profile,
	}, nil
}
